### DISCUSSION TITLE
关于ListBox的ItemsPanelTemplate的问题

### DISCUSSION THREAD

**Discussion by Zhaozhuofan** [BYJRK/DotNet-Discussions#258] [WPF 开发]
UP您好,我想把ListBox作成下图的样式
<img width="531" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/103978682/d5e60b3a-52c5-4ff7-9332-39306aabf2a3">
但我在写ListBox的ItemPanelTemplate时,vs提示我_无法显式修改Panel的Children集合,它作用于ItemsControl的ItemsPanel.ItemsControl生成Paneld的子元素_
所以想请教一下您有没有能把ListBox变成这样两排的好方法?

**Comment by lmccccl** [comment]
自定义容器

**Comment by BYJRK** [comment]
这个恐怕没有什么简单的办法，因为简单来看，外面可以是一个 StackPanel，而其中每一行都是一个水平居中的横向 StackPanel，并且数量似乎是不确定的（比如可定制，可变，或根据容器的宽度来决定等）。纯 XAML 写 Template 恐怕无法实现这么复杂的逻辑。你可能需要借助附加属性、行为，或者自定义控件了

**Reply by BYJRK** [reply]
不过这个界面挺有意思的。是想做一个什么APP呢？有没有开源代码

**Reply by Zhaozhuofan** [reply]
https://github.com/Zhaozhuofan/konosuba.git
学习了您原神的视频之后做的,模仿的是下面的那个网页
http://konosuba.com/1st/character/
还想请教您一下就是我再写MarkUpExtension的时候在大括号里用TAB键跳不出大括号之外,在网上查了好久也没有答案,
想问一下您时用的什么快捷键跳出的大括号?

**Reply by BYJRK** [reply]
我一时没想起来实用MarkupExtension的情形。印象中好像不能跳出吧？我一般的做法可能是按住Ctrl键然后用方向键，快速跳过单词

**Reply by BYJRK** [reply]
另外我看了一下，既然konosuba总共也就这么几个角色，要不然就直接指定每行子元素的数量吧，不用搞那么复杂了

**Reply by Zhaozhuofan** [reply]
谢谢您的回答,但我还想问一个问题.
现在我的每个人物的头像都是圆形的的,但可点击范围却是正方形.
这种情况需要修改ListBox的哪个属性呢?  我试着改了很多都没有成功...
希望您能在指点指点

**Reply by Zhaozhuofan** [reply]
是我傻了,我的图片本来就是正方形,只不过除了圆形的部分外其他的部分透明而已...
<img width="97" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/103978682/8ea80487-d504-4818-b22c-0a4427ed491f">


<img width="549" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/103978682/be456a3c-94a2-44b0-923b-7af7fe1ba80d">

我把ItemTemplate改成这个感觉之后虽然点击范围的问题解决了,但又出现了新的问题...
就是第一次进入程序后什么listbox也不显示,必须点击一次后,listbox才会正常显示出来.
希望您能指点我一下这是为什么

**Reply by BYJRK** [reply]
你点击了什么，以及触发了什么，才会让图片正常显示出来呢？

**Reply by Zhaozhuofan** [reply]
程序进去后是这个样子
<img width="777" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/103978682/dcabb062-ff71-4f62-bf5e-538521a1b92c">

之后点击头像对应的位置后才会出现头像的图片
<img width="790" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/103978682/b2c0bffd-f61a-4fd6-b6cc-c00aeb19f524">

使用Image控件绑定时并没有这个问题,程序进去后人物头像会自己加载出来,但是我想修改点击范围于是就使用了Ellipse控件,
可是设置了ImageBrush的ImageSource的Binding后就发现了该问题,希望Up您能给指点指点

**Reply by hinadesu39** [reply]
一张方的图片可以直接通过设置image的clip属性把他裁剪成圆的，没必要用ellipse吧

**Reply by Zhaozhuofan** [reply]
谢谢指导
我重写了ListBox的ItemContainerStyle,把grid改成了圆形后解决了
<img width="846" alt="image" src="https://github.com/BYJRK/DotNet-Discussions/assets/103978682/cf1d1f7c-3422-4efc-952b-2c812c66c67e">

谢谢你的回复

**Comment by SlimeNull** [comment]
这就不得不推一下我的 WPF Suite 库了。你的布局需求是，自动换行+居中，所以，用FlexPanel 可以轻松解决。

在 nuget 安装 EleCho.WpfSuite
引入命名空间 xmlns:ws="https://github.com/OrgEleCho/EleCho.WpfSuite"

将 ListBox 的 ItemsPanelTemplate 设置为 ws:FlexPanel
并且设定属性：
1. MainAlignment="Center" (在主轴上居中对齐，默认主轴是水平方向）
2. Wrap="Wrap" (允许换行）

如果你希望在各个元素之间流出间距，也可以设定以下属性：
1. MainSpacing="8" (主轴上元素间距设为8）
2. CrossSpacing="8" (交叉轴上间距设为8）

另外，WPF Suite 还提供了支持圆角的 Image 控件，你也可以试试用。
